src/lib.rs

```diff
@@ -1,25 +1,75 @@
 #![no_std]
-pub mod contracts;
 extern crate alloc;
 #[cfg(test)]
 mod e2e;
+#[cfg(test)]
+mod test_helpers;
+
 
 use gstd::{
     msg::{self, reply},
     prelude::*,
 };
 use io::*;
-
+use contracts::{errors::InvariantError, FeeTier, FeeTiers};
+use math::percentage::Percentage;
+use decimal::*;
 #[derive(Default, Clone)]
 pub struct Invariant {
     pub config: InvariantConfig,
+    pub fee_tiers: FeeTiers,
 }
 
 impl Invariant {
-    pub fn change_protocol_fee(&mut self, protocol_fee: u128) {
+    pub fn change_protocol_fee(&mut self, protocol_fee: u128) -> Result<u128, InvariantError> {
+        if !self.caller_is_admin() {
+            return Err(InvariantError::NotAdmin);
+        }
+
         self.config.protocol_fee = protocol_fee;
+        Ok(self.config.protocol_fee)
+    }
+
+    pub fn get_protocol_fee(&self) -> u128 {
+        self.config.protocol_fee
+    }
+
+    pub fn add_fee_tier(&mut self, fee_tier: FeeTier) -> Result<FeeTier, InvariantError> {
+        if fee_tier.tick_spacing == 0 || fee_tier.tick_spacing > 100 {
+            return Err(InvariantError::InvalidTickSpacing);
+        }
+
+        if fee_tier.fee >= Percentage::from_integer(1) {
+            return Err(InvariantError::InvalidFee);
+        }
+
+        if !self.caller_is_admin() {
+            return Err(InvariantError::NotAdmin);
+        }
+
+        self.fee_tiers.add(fee_tier)?;
+        Ok(fee_tier)
+    }
+
+    pub fn fee_tier_exists(&self, fee_tier: FeeTier) -> bool {
+        self.fee_tiers.contains(fee_tier)
+    }
+
+    pub fn remove_fee_tier(&mut self, fee_tier: FeeTier) -> Result<FeeTier, InvariantError> {
+        if !self.caller_is_admin() {
+            return Err(InvariantError::NotAdmin);
+        }
+
+        self.fee_tiers.remove(fee_tier)?;
+        Ok(fee_tier)
+    }
 
-        reply(InvariantEvent::ProtocolFeeChanged(protocol_fee), 0).expect("Unable to reply");
+    pub fn get_fee_tiers(&self) -> Vec<FeeTier> {
+        self.fee_tiers.get_all()
+    }
+
+    fn caller_is_admin(&self) -> bool {
+        msg::source() == self.config.admin
     }
 }
 
```

Comment:
Is there a possibility to use lazy initialization to avoid wrapping the global state in an `Option`?